File: clients/go/model/debug_state.go
package model

import "io"
import . "komus24/stream"

// TODO - Document
type DebugState struct {
    // TODO - Document
    PressedKeys []string
}

func NewDebugState(pressedKeys []string) DebugState {
    return DebugState {
        PressedKeys: pressedKeys,
    }
}

// Read DebugState from reader
func ReadDebugState(reader io.Reader) DebugState {
    var pressedKeys []string
    pressedKeys = make([]string, ReadInt32(reader))
    for pressedKeysIndex := range pressedKeys {
        var pressedKeysElement string
        pressedKeysElement = ReadString(reader)
        pressedKeys[pressedKeysIndex] = pressedKeysElement
    }
    return DebugState {
        PressedKeys: pressedKeys,
    }
}

// Write DebugState to writer
func (debugState DebugState) Write(writer io.Writer) {
    pressedKeys := debugState.PressedKeys
    WriteInt32(writer, int32(len(pressedKeys)))
    for _, pressedKeysElement := range pressedKeys {
        WriteString(writer, pressedKeysElement)
    }
}

// Get string representation of DebugState
func (debugState DebugState) String() string {
    stringResult := "{ "
    stringResult += "PressedKeys: "
    pressedKeys := debugState.PressedKeys
    stringResult += "[ "
    for pressedKeysIndex, pressedKeysElement := range pressedKeys {
        if pressedKeysIndex != 0 {
            stringResult += ", "
        }
        stringResult += "\"" + pressedKeysElement + "\""
    }
    stringResult += " ]"
    stringResult += " }"
    return stringResult
}

File: clients/php/Model/DebugState.php
<?php

namespace Model {
    require_once 'Stream.php';

    /**
     * TODO - Document
     */
    class DebugState
    {
        /**
         * TODO - Document
         */
        public array $pressedKeys;
    
        function __construct(array $pressedKeys)
        {
            $this->pressedKeys = $pressedKeys;
        }
    
        /**
         * Read DebugState from input stream
         */
        public static function readFrom(\InputStream $stream): DebugState
        {
            $pressedKeys = [];
            $pressedKeysSize = $stream->readInt32();
            for ($pressedKeysIndex = 0; $pressedKeysIndex < $pressedKeysSize; $pressedKeysIndex++) {
                $pressedKeysElement = $stream->readString();
                $pressedKeys[] = $pressedKeysElement;
            }
            return new DebugState($pressedKeys);
        }
        
        /**
         * Write DebugState to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(count($this->pressedKeys));
            foreach ($this->pressedKeys as $element) {
                $stream->writeString($element);
            }
        }
    }
}